### PR TITLE
Fix - TLS hnadshake failure for oci pull (artifact repo).

### PR DESCRIPTION
**Pull request by suryatech27-cloud**
**What this PR does / why we need it**:
Fix for https://github.com/helm/helm/issues/10597
- Added 2-way TLS Support for oci pull for artifact repository which causes TLS handshake failure error.

**Special notes for your reviewer**:
- For now, this fix will be going to support Linux only.
**If applicable**:
- [ ] this PR contains documentation
- [ ] this PR contains unit tests
- [ ] this PR has been tested for backwards compatibility

### CODE REVIEW

**Comment by bacongobbler** [comment]
Hi @suryatech27-cloud, generally speaking we ask that pull requests come with a set of unit tests to ensure the feature works as expected. Can you please have a look at existing test cases and see if you can write one for this?

Additionally, is there a way maintainers should test this PR?

Thank you.

File: pkg/registry/client.go [hunk 1]
```diff
@@ -131,6 +135,94 @@ func NewClient(options ...ClientOption) (*Client, error) {
 	return client, nil
 }
 
+func NewCrosClient(chartref string, options ...ClientOption) (*Client, error) {
```
Comment (bacongobbler) [inline pkg/registry/client.go:138]: what is a "Cros" client?

File: pkg/action/pull.go [hunk 2]
```diff
@@ -125,7 +125,26 @@ func (p *Pull) Run(chartRef string) (string, error) {
 
 	saved, v, err := c.DownloadTo(chartRef, p.Version, dest)
 	if err != nil {
-		return out.String(), err
+		if registry.IsOCI(chartRef) && strings.Contains(fmt.Sprint(err), "remote error: tls: handshake failure") {
+			registryClient, err := registry.NewCrosClient(chartRef,
```
Comment (bacongobbler) [inline pkg/action/pull.go:129]: Why are we re-creating a one-off registry client instead of altering `c.RegistryClient`? Shouldn't this setup code be part of [NewOciGetter](https://github.com/helm/helm/blob/c137bfb68b4f8b7a10687f043b0dbccb09574969/pkg/getter/ocigetter.go#L65)?

**Review by bacongobbler — CHANGES_REQUESTED** [review]

File: pkg/registry/client.go (hunk 1 above)
Comment (suryatech27-cloud) [reply pkg/registry/client.go:138]: **Cros  (Connection Roundtripper Over Ssl)** 
This **NewCrosclient** only uses two-way TLS support. Due to less experience in golang development, I thought maybe my code affect existing functionality so I created a new function named as newCrosClient.

If that naming convention does not match helm expectations we can change that.

File: pkg/action/pull.go (hunk 2 above)
Comment (suryatech27-cloud) [reply pkg/action/pull.go:129]: Yes, you are right but there is one problem if we add that setup code in [NewOciGetter](https://github.com/helm/helm/blob/c137bfb68b4f8b7a10687f043b0dbccb09574969/pkg/getter/ocigetter.go#L65) then we need to distinguish between one-way TLS and two-way TLS. 
if we validate then a specific client setup will be done because we don't need a two-way tls client setup for one-way and vice-versa.

`if registry.IsOCI(chartRef) && strings.Contains(fmt.Sprint(err), "remote error: tls: handshake failure")`
so above code is validating when client couldn't establish connection(required client cert and key) then we setting up the client with roundtripper/tlsconfig data.

Comment (suryatech27-cloud) [reply pkg/action/pull.go:129]: **(When the Registry server is a one-way TLS that time it works fine  but when the registry server is a two-way TLS , the server looking for client certificate/key (two-way TLS), then the helm client does not provide the client info (TLSconfig) that causes a handshake failure. so my code is to resolve those issue)**

**Comment by suryatech27-cloud** [comment]
> Hi @suryatech27-cloud, generally speaking we ask that pull requests come with a set of unit tests to ensure the feature works as expected. Can you please have a look at existing test cases and see if you can write one for this?
> 
> Additionally, is there a way maintainers should test this PR?
> 
> Thank you.

@bacongobbler, Yes there is a way to test this PR, just need to enable two-way SSL/TLS in the registry server and validate.

File: internal/tlsutil/cfg.go [hunk 3]
```diff
@@ -56,3 +62,77 @@ func ClientConfig(opts Options) (cfg *tls.Config, err error) {
 	cfg = &tls.Config{InsecureSkipVerify: opts.InsecureSkipVerify, Certificates: []tls.Certificate{*cert}, RootCAs: pool}
 	return cfg, nil
 }
+
+func ReadCertFromSecDir(host string) (opts Options, err error) {
+	if runtime.GOOS == "windows" || runtime.GOOS == "unix" {
+		fmt.Printf("%v OS not supported for this oci pull. Contact your administrator for more information !!!", runtime.GOOS)
```
Comment (bacongobbler) [inline internal/tlsutil/cfg.go:68]: Why should the user have to contact their administrator for Helm support?

File: internal/tlsutil/cfg.go [hunk 4]
```diff
@@ -56,3 +62,77 @@ func ClientConfig(opts Options) (cfg *tls.Config, err error) {
...
+		os.Exit(1)
```
Comment (bacongobbler) [inline internal/tlsutil/cfg.go:69]: https://pkg.go.dev/log#Fatalf

File: internal/tlsutil/cfg.go [hunk 5]
```diff
@@ -56,3 +62,77 @@ func ClientConfig(opts Options) (cfg *tls.Config, err error) {
...
+	} else {
+		cmd, err := exec.Command("helm", "env", "HELM_CLIENT_TLS_CERT_DIR").Output()
+		if err != nil {
+			fmt.Printf("Error : %s", err)
```
Comment (bacongobbler) [inline internal/tlsutil/cfg.go:73]: https://pkg.go.dev/log#Fatalf

File: internal/tlsutil/cfg.go [hunk 6]
```diff
@@ -56,3 +62,77 @@ func ClientConfig(opts Options) (cfg *tls.Config, err error) {
...
+
+		if clientCertDir[len(clientCertDir)-1] != '/' {
+			clientCertDir = fmt.Sprintf("%s/%s", clientCertDir, host)
+		} else {
+			clientCertDir = fmt.Sprintf("%s%s", clientCertDir, host)
+		}
+		if _, err := os.Stat(clientCertDir); err != nil {
+			if os.IsNotExist(err) {
+				os.MkdirAll(clientCertDir, os.ModePerm)
+				return opts, errors.Wrapf(err, clientCertDir, "%v\n%v Directory created.")
```
Comment (bacongobbler) [inline internal/tlsutil/cfg.go:90]: why are we creating the directory, then throwing back the same error?

File: internal/tlsutil/cfg.go [hunk 7]
```diff
@@ -56,3 +62,77 @@ func ClientConfig(opts Options) (cfg *tls.Config, err error) {
...
+			}
+		} else {
+
+			if files, err := ioutil.ReadDir(clientCertDir); err == nil {
+				for _, file := range files {
+					if filepath.Ext(file.Name()) == ".crt" {
```
Comment (bacongobbler) [inline internal/tlsutil/cfg.go:96]: why is this necessary?

File: internal/tlsutil/cfg.go [hunk 8]
```diff
@@ -56,3 +62,77 @@ func ClientConfig(opts Options) (cfg *tls.Config, err error) {
...
+						opts.CaCertFile = fmt.Sprintf("%s/%s", clientCertDir, file.Name())
+					} else if filepath.Ext(file.Name()) == ".pem" {
+						opts.CertFile = fmt.Sprintf("%s/%s", clientCertDir, file.Name())
+					} else if filepath.Ext(file.Name()) == ".key" {
+						opts.KeyFile = fmt.Sprintf("%s/%s", clientCertDir, file.Name())
+					}
+				}
+			} else {
+				fmt.Printf(" Certificate not found in current directory - %v\n ", err)
```
Comment (bacongobbler) [inline internal/tlsutil/cfg.go:105]: https://pkg.go.dev/log#Fatalf

**Review by bacongobbler — CHANGES_REQUESTED** [review]

**Comment by bacongobbler** [comment]
> Yes there is a way to test this PR, just need to enable two-way SSL/TLS in the registry server and validate.

I do not know how to set that up. Please provide a set of instructions so developers can test this PR. It is your responsibility to teach others how to test this.

File: pkg/action/pull.go (hunk 2 above)
Comment (bacongobbler) [reply pkg/action/pull.go:129]: Just make two-way TLS an `Option` like we do for the rest of the getter feature flags.

for example:

https://github.com/helm/helm/blob/c137bfb68b4f8b7a10687f043b0dbccb09574969/pkg/getter/getter.go#L54

**Comment by bacongobbler** [comment]
Alternatively provide support for two-way TLS with the test server and use that against these changes.

https://github.com/helm/helm/blob/main/pkg/repo/repotest/server.go#L75

File: internal/tlsutil/cfg.go (hunk 3 above)
Comment (suryatech27-cloud) [reply internal/tlsutil/cfg.go:68]: Yeah, you are right user don't need to contact their administrator because our user message is sufficient.
I will remove that message.
Thanks.

File: internal/tlsutil/cfg.go (hunk 6 above)
Comment (suryatech27-cloud) [reply internal/tlsutil/cfg.go:90]: Correct, that has no meaning if we are creating the directory and throwing back the same error.
I will remove the directory creation code and will update the error message.

Thanks.

**Comment by bacongobbler** [comment]
I just realized this pull request is targeting the release-3.8 branch. In the future, please target your pull requests against the `main` branch.

Closing as we are not going to merge PRs against a release branch. Please re-open against `main`. Thanks.

**Comment by yxxhero** [comment]
@suryatech27-cloud ping
